### PR TITLE
SUIT CI tests DON'T MERGE

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds CI testing to examples/suit_update.
Based on https://github.com/future-proof-iot/RIOT/pull/8.

The PR includes some dependencies already PR'ed to master, so it should not be merged.

The test script requires the flashed APP_VER to be 0 before starting. This might break the examples, and needs to be checked.

Other than that, "make test" using sudo start_network.sh doesn't work at all for me. @fjmolinas I guess it did for you?

### CODE REVIEW

**Comment by kaspar030** [comment]
> The test script requires the flashed APP_VER to be 0 before starting. This might break the examples, and needs to be checked.

This I don't know yet how to work around. CI and test script need APP_VER=0. For users, APP_VER=$epoch is nicest. Would be a shame to have an extra test application just to acomodate this difference.

**Comment by kaspar030** [comment]
The branch contains some newer master commits and some master PR dependencies. Only the two topmost commits should be reviewed here.

File: examples/suit_update/tests/01-run.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    print("TEST PASSED")
+
+
+if __name__ == "__main__":
+    try:
+        res = 1
+        tmpdir, aiocoap_process = start_aiocoap_fileserver()
+        # TODO: wait for coap port to be available
+
+        publish(tmpdir.name, COAP_HOST, 1)
```
Comment (fjmolinas) [inline examples/suit_update/tests/01-run.py]:
```
publish(tmpdir.name, COAP_HOST, 1, 1)
publish(tmpdir.name, COAP_HOST, 2, 2)
```

File: examples/suit_update/Makefile [hunk 2]
```diff
@@ -39,61 +39,77 @@ USEMODULE += shell_commands
...
+  endif
 endif
-endif
 
-# riotboot is currently incompatible with the softdevice.
-# that gets automatically selected with gnrc_netdev_default, unless nrfmin is used.
+# The test script assumes initially version "0" is flashed,
+# as done by the CI.
+# Thus default to that (instead of epoch set by makefiles/boot/riotboot.inc.mk).
+APP_VER ?= 0
```
Comment (fjmolinas) [inline examples/suit_update/Makefile]: Could be removed, see above comment.

File: examples/suit_update/tests/01-run.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+MANIFEST_TIMEOUT = 15
+
+USE_ETHOS = int(os.getenv("USE_ETHOS", "1"))
+TAP = os.getenv("TAP", "tap0")
+
+
+def start_aiocoap_fileserver():
+    tmpdirname = tempfile.TemporaryDirectory()
+    aiocoap_process = subprocess.Popen(
+        "exec aiocoap-fileserver %s" % tmpdirname.name, shell=True
```
Comment (fjmolinas) [inline examples/suit_update/tests/01-run.py:29]: It needs to be documented that this was added to your path, I thinks that what you did right?

File: examples/suit_update/tests/01-run.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    )
+
+    return tmpdirname, aiocoap_process
+
+
+def cleanup(tmpdir, aiocoap_process):
+    aiocoap_process.kill()
```
Comment (fjmolinas) [inline examples/suit_update/tests/01-run.py:36]: This doesn't kill everything properly, specificity fails to clean up the created `tap` interface. SIGKILL is killing all child processes but not removing the interface, this is because in the script:

`trap "cleanup" INT QUIT TERM EXIT`.

Can you tell if there is a reason for this? Or should we push the change in master? I'll take a look at the history.

**Review by fjmolinas — CHANGES_REQUESTED** [review]
What does it have to do with RIOT-OS#31 "migrate native to current Makefile structure" ?

> Other than that, "make test" using sudo start_network.sh doesn't work at all for me. @fjmolinas I guess it did for you?

Are you doing `sudo make test`? Or running it separately?

Anyway this has changes my setup a lot, publishing is done in the same go so it is running with sudo and publishing files with sudo. This means everything needs to be killed with sudo too, and `process.kill()` doesn't cleanup everything `start_network.sh` sets up, so I usually have to call test twice, once for cleanup and another to setup and tun.

> This I don't know yet how to work around. CI and test script need APP_VER=0. For users, APP_VER=$epoch is nicest. Would be a shame to have an extra test application just to accommodate this difference.

Do you really need `APP_VER=0`? I bodged a workaround, a gist [HERE](https://gist.github.com/fjmolinas/ba517d4407905e3026fe5bbb0b99719d).

Anyway it works for me with or without the patch but only for the first round of update, it is always using "latest" so the second try fails. When publishing it did specify a `latest_name` to easily pick the fw to notify, if you change the publish to:

```
        publish(tmpdir.name, COAP_HOST, 1, 1)
        publish(tmpdir.name, COAP_HOST, 2, 2)
```

and `SUIT_NOTIFY_VERSION` to `SUIT_MANIFEST_SIGNED_LATEST` for notifying it works. `SUIT_NOTIFY_VERSION` is not in the makefile so does nothing, you could also set `SUIT_NOTIFY_VERSION ?= SUIT_MANIFEST_SIGNED_LATEST' and change the notify recipe to:

```
suit/notify: | $(filter suit/publish, $(MAKECMDGOALS))
	@test -n "$(SUIT_CLIENT)" || { echo "error: SUIT_CLIENT unset!"; false; }
	aiocoap-client -m POST "coap://$(SUIT_CLIENT)/suit/trigger" \
		--payload "$(SUIT_COAP_ROOT)/$$(basename $(SUIT_NOTIFY_VERSION))" && \
		echo "Triggered $(SUIT_CLIENT) to update."
```

Doing this changes the test works for me. Here is my diff:

```
index 2b5986853..3be2de5f4 100755
--- a/examples/suit_update/tests/01-run.py
+++ b/examples/suit_update/tests/01-run.py
@@ -45,7 +45,7 @@ def notify(coap_server, client_url, version=None):
         "SUIT_CLIENT={}".format(client_url),
     ]
     if version is not None:
-        cmd.append("SUIT_NOTIFY_VERSION={}".format(version))
+        cmd.append("SUIT_MANIFEST_SIGNED_LATEST={}".format(version))
     assert not subprocess.call(cmd)
 
 
@@ -111,8 +111,8 @@ if __name__ == "__main__":
         tmpdir, aiocoap_process = start_aiocoap_fileserver()
         # TODO: wait for coap port to be available
 
-        publish(tmpdir.name, COAP_HOST, 1)
-        publish(tmpdir.name, COAP_HOST, 2)
+        publish(tmpdir.name, COAP_HOST, 1, 1)
+        publish(tmpdir.name, COAP_HOST, 2, 2)
 
```

And the command sequence I ran is:

```
make -C examples/suit_update/ BOARD=samr21-xpro -j3 flash

sudo make -C examples/suit_update/ BOARD=samr21-xpro test
```

PS: small comment. I rather have Francisco Molina as naming in the commit co-authorship (and for all future cases).

**Comment by kaspar030** [comment]
> What does it have to do with [RIOT-OS#31](https://github.com/RIOT-OS/RIOT/pull/31) "migrate native to current Makefile structure" ?

Messup, I meant your PR: future-proof-iot/RIOT#8

**Comment by kaspar030** [comment]
>  small comment. I rather have Francisco Molina as naming in the commit co-authorship (and for all future cases).

fixed

**Comment by fjmolinas** [comment]
@kaspar030 I see you pushed a lot of changes, should I re-test? or are more changes coming?

**Comment by kaspar030** [comment]
> @kaspar030 I see you pushed a lot of changes, should I re-test? or are more changes coming?

Yes, please re-test! I'll maybe be pushing documentation fixes.

File: examples/suit_update/README.md [hunk 5]
```diff
@@ -529,3 +510,23 @@ suit/genkey: this recipe generates a ed25519 key to sign the manifest
...
+## Automatic test
+[Automatic test]: #test
+
+This applications ships with an automatic test. To run it
+
+- ensure the [prerequisites] are installed
+
+- compile the application: 
+
+    $ make -Cexamples/suit_update make clean all flash -j8
```
Comment (fjmolinas) [inline examples/suit_update/README.md]: wow wow wow -j8 seems like a bit much, missing space in `-Cexamples/suit_update `

**Comment by kaspar030** [comment]
I pushed a small for for the Makefile, "$SUIT_KEY" was used instead of "$SUIT_SEC", in "TEST_FILES_EXTRA". Doesn't concern local testing, only "make test-murdock" was broken.

File: examples/suit_update/README.md (hunk 5 above)
Comment (fjmolinas) [reply examples/suit_update/README.md]: also `twice in the command`, this should do:

`$ make -C examples/suit_update clean all flash -j`

**Comment by aabadie** [comment]
What is this PR about exactly ? Does it contain the branch that will actually be PRed upstream ?

If not, I suggest opening small PRs against suit-dev so we can review them. I don't like last minute changes like this. Even if they are *just* documentation fixes.

File: dist/tools/ethos/setup_network.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+start_uhcpd() {
+    ${UHCPD} ${TAP} ${PREFIX}
+}
+
+TAP=$1
+PREFIX=$2
+_USER=$3
+UHCPD="../uhcpd/bin/uhcpd"
```
Comment (fjmolinas) [inline dist/tools/ethos/setup_network.sh]: This does not work if its not called from `/dist/tools/ethos/`, change for `UHCPD="$(readlink -f $(dirname $0)"/../uhcpd/bin")/uhcpd"`?

**Review by fjmolinas — CHANGES_REQUESTED** [review]
Some minor comments, otherwise works like a charm and I like the approach.

File: examples/suit_update/README.md [hunk 7]
```diff
@@ -167,6 +168,15 @@ In this case the RIOT node can be reached from the host using its global address
 `aiocoap-fileserver` is used for hosting firmwares available for updates. Device
 retrieve the new firmware using the CoAP protocol.
 
+*either*:
```
Comment (aabadie) [inline examples/suit_update/README.md]: This seems unrelated to the goal of this PR, right ?

File: LOSTANDFOUND.md [hunk 8]
```diff
@@ -100,3 +100,14 @@ Author(s):
 Reason for removal:
 - code broken and excluded from all tests for a long time
 - no maintainer available
+
+### gnrc_pktbuf_duplicate_upto() [b83430aa625a1d42f11f9badf5e5cfbb8efacd99]
```
Comment (aabadie) [inline LOSTANDFOUND.md]: Looks like there's some rebase issue with this PR.

File: examples/suit_update/README.md (hunk 7 above)
Comment (fjmolinas) [reply examples/suit_update/README.md]: No, the lower ~~commit~~edits say fileserver needs to be added it to $PATH and this is then used in the test script to setup the fileserver.

Comment (aabadie) [reply examples/suit_update/README.md]: How is this useful from the user point of view ? except by introducing confusion ? I mean this way of installing is only done for the CI so there's no need to document it here. Or maybe only give one way of installing the fileserver ?

Comment (fjmolinas) [reply examples/suit_update/README.md]: Hmm you are right, maybe it should be split in test documentation?

File: examples/suit_update/README.md (hunk 5 above)
Comment (kaspar030) [reply examples/suit_update/README.md]: > wow wow wow -j8 seems like a bit much, missing space in `-Cexamples/suit_update `

fixed, very used to type "-j8" on my laptop... space is not necessary, i never use it. anyhow, added them where missing for consistency.

Comment (kaspar030) [reply examples/suit_update/README.md]: > also `twice in the command`, this should do:

removed the extra make

File: dist/tools/ethos/setup_network.sh (hunk 6 above)
Comment (kaspar030) [reply dist/tools/ethos/setup_network.sh]: good point. changed to ```: ${UHCPD:="$(readlink -f $(dirname $0)"/../uhcpd/bin")/uhcpd"}``` so it can still be overridden.

File: examples/suit_update/README.md (hunk 7 above)
Comment (kaspar030) [reply examples/suit_update/README.md]: I changed the prerequisites section to include the wget method, saving to ```~/.local/bin```.
Also added a paragraph with the requirement to add "~/.local/bin" to PATH.

File: LOSTANDFOUND.md (hunk 8 above)
Comment (kaspar030) [reply LOSTANDFOUND.md]: it is just that the PR shows the new commits in master that are not yet in suit-dev

**Comment by kaspar030** [comment]
> What is this PR about exactly ? Does it contain the branch that will actually be PRed upstream ?
 
This is my WIP branch that integrates suit-dev and the required upstream PRs, and adds the automatic test on top.

> If not, I suggest opening small PRs against suit-dev so we can review them. 

Not possible, as suit-dev is missing dependencies.

File: examples/suit_update/tests/01-run.py (hunk 3 above)
Comment (kaspar030) [reply examples/suit_update/tests/01-run.py:29]: yup

File: examples/suit_update/tests/01-run.py (hunk 4 above)
Comment (kaspar030) [reply examples/suit_update/tests/01-run.py:36]: Ah, the shell script cannot trap "KILL". Need to change to send "TERM" first, then "KILL" a little later...

Comment (kaspar030) [reply examples/suit_update/tests/01-run.py:36]: sorry, this was outdated, with setup_network.sh, this isn't an issue anymore

**Comment by kaspar030** [comment]
- rebased all branches
- fixed uhcp output messing up the slot printout by guarding with irq_disable()
- addressed @fjmolinas comments

I propose I create the squashed master-PR branch, we all give it a last look, and if everyone's happy, I open the PR?

**Comment by fjmolinas** [comment]
> I propose I create the squashed master-PR branch, we all give it a last look, and if everyone's happy, I open the PR?

Lets do that, are you ok with following the procedure that I sent on the mailing list?

**Comment by kaspar030** [comment]
> Lets do that, are you ok with following the procedure that I sent on the mailing list?

didn't get that, when did you send it?

**Comment by kaspar030** [comment]
> didn't get that, when did you send it?

ah got it, sorry

**Comment by fjmolinas** [comment]
Should we close this one?
